lightning/src/ln/outbound_payment.rs

```diff
@@ -1768,6 +1787,74 @@ impl OutboundPayments {
 	}
 }
 
+fn set_max_path_length(
+	route_params: &mut RouteParameters, recipient_onion: &RecipientOnionFields,
+	keysend_preimage: Option<PaymentPreimage>, best_block_height: u32,
+) -> Result<(), ()> {
+	const PAYLOAD_HMAC_LEN: usize = 32;
+	const UNBLINDED_INTERMED_PAYLOAD_LEN: usize = PAYLOAD_HMAC_LEN
+		+ 1 + 1 + 8 // amt_to_forward
+		+ 1 + 1 + 4 // outgoing_cltv
+		+ 1 + 1 + 8; // short_channel_id
+
+	let num_reserved_bytes = match &route_params.payment_params.payee {
+		Payee::Blinded { route_hints, .. } => {
+			let (blinded_payinfo, largest_path) = route_hints
+				.iter()
+				.max_by(|(_, path_a), (_, path_b)|
+					path_a.serialized_length().cmp(&path_b.serialized_length()))
+				.ok_or(())?;
+
+			largest_path.blinded_hops
+				.iter()
+				.rev()
+				.skip(1)
+				.map(|bh| msgs::OutboundOnionPayload::BlindedForward {
+					encrypted_tlvs: &bh.encrypted_payload,
+					// For simplicity, always set the intro_node_blinding_point in the
+					// final payload.
+					intro_node_blinding_point: None,
+				})
+				.chain(core::iter::once(msgs::OutboundOnionPayload::BlindedReceive {
+					sender_intended_htlc_amt_msat: route_params.final_value_msat,
+					total_msat: route_params.final_value_msat,
+					cltv_expiry_height:
+						best_block_height.saturating_add(blinded_payinfo.cltv_expiry_delta as u32),
+					encrypted_tlvs: largest_path.blinded_hops
+						.last()
+						.map(|bh| &bh.encrypted_payload)
+						.unwrap_or(&Vec::new()),
+					intro_node_blinding_point: Some(largest_path.blinding_point),
+					keysend_preimage,
+					custom_tlvs: &recipient_onion.custom_tlvs,
+				}))
+				.map(|payload| payload.serialized_length().saturating_add(PAYLOAD_HMAC_LEN))
+				.fold(0, |acc: usize, payload_len| acc.saturating_add(payload_len))
+		},
+		Payee::Clear { final_cltv_expiry_delta, .. } => {
+			msgs::OutboundOnionPayload::Receive {
+				payment_data: recipient_onion.payment_secret.map(|payment_secret| {
+					msgs::FinalOnionHopData { payment_secret, total_msat: route_params.final_value_msat }
+				}),
+				payment_metadata: recipient_onion.payment_metadata.as_ref(),
+				keysend_preimage,
+				custom_tlvs: &recipient_onion.custom_tlvs,
+				sender_intended_htlc_amt_msat: route_params.final_value_msat,
+				cltv_expiry_height: best_block_height.saturating_add(*final_cltv_expiry_delta),
+			}
```

Comment:
Definitely agree that that's cleaner. Do you think it's worth trying to avoid the `Path::hops`/`Path::blinded_tail` allocation by passing in an iterator of [blinded_]hops as well? Anyway I'll take a look.